[Wrangler] Explain bundling & rules

penalosa opened the pull request:
Explain how Wrangler bundles, and when it _doesn't_ bundle

GregBrimble commented on content/workers/wrangler/bundling.md:
````diff
@@ -0,0 +1,48 @@
+This is also the basis of WASM support with Wrangler. To use a WASM module in a Worker developed with Wrangler, add the following to your Worker:
+
+```js
+import wasm from "example.wasm"; // Where `example.wasm` is a file in your local directory
````
```suggestion
import wasm from "./example.wasm"; // Where `example.wasm` is a file in your local directory
```

deadlypants1973 commented on content/workers/wrangler/bundling.md:
```diff
@@ -0,0 +1,48 @@
+
+By default, Wrangler bundles your Worker code using [`esbuild`](https://esbuild.github.io/). This means that Wrangler has built in support for importing modules from [npm](https://www.npmjs.com/) defined in your `package.json`. If you'd like to see the exact code that Wrangler will upload to Cloudflare, you can run `wrangler publish --dry-run --outdir dist`, which will show your Worker code post  ...
+
+## Files which won't be bundled
```
```suggestion
## Files which will not be bundled
```

deadlypants1973 commented on content/workers/wrangler/bundling.md:
```diff
@@ -0,0 +1,48 @@
+- `.bin`
+- `.wasm`
+
+See the [rules configuration](/workers/wrangler/configuration/#bundling) to customise these file types.
```
```suggestion
Refer to [Bundling configuration](/workers/wrangler/configuration/#bundling) to customize these file types.
```

deadlypants1973 commented on content/workers/wrangler/bundling.md:
```diff
@@ -0,0 +1,48 @@
+
+## Files which won't be bundled
+
+The effect of bundling your Worker code is to take multiple modules and bundle them into one. Sometimes, however, that isn't the behaviour you want. There might be modules that _shouldn't_ be inlined directly into the bundle. For instance, bundling a WASM file into your Javascript Worker wouldn't make a whole lot of sense—that's something you want to upload as a separate module that can be import ...
```
```suggestion
Bundling your Worker code is taking multiple modules and bundling them into one. Sometimes, that is not the desired behavior. There might be modules that should not be inlined directly into the bundle. For example, instead of bundling a WASM file into your Javascript Worker, you would want to upload the WASM file as a separate module that can be imported at runtime. Wrangler supports this for the following file types:
```

deadlypants1973 commented on content/workers/wrangler/bundling.md:
```diff
@@ -0,0 +1,48 @@
+
+See the [rules configuration](/workers/wrangler/configuration/#bundling) to customise these file types.
+
+This means that for an import like the following:
```
```suggestion
Review the following import:
```

deadlypants1973 commented on content/workers/wrangler/bundling.md:
````diff
@@ -0,0 +1,48 @@
+import data from "./example.html"; // Where `example.html` is a file in your local directory
+```
+
+the variable `data` will be a string containing the contents of `example.html`.
````
```suggestion
In the above import, the variable `data` will be a string containing the contents of `example.html`.
```

deadlypants1973 commented on content/workers/wrangler/bundling.md:
````diff
@@ -0,0 +1,48 @@
+```
+
+{{<Aside type="warning">}}
+Cloudflare Workers doesn't support `WebAssembly.instantiateStreaming()`
````
```suggestion
Cloudflare Workers does not support `WebAssembly.instantiateStreaming()`
```

deadlypants1973 commented on content/workers/wrangler/bundling.md:
```diff
@@ -0,0 +1,48 @@
+
+the variable `data` will be a string containing the contents of `example.html`.
+
+This is also the basis of Wasm support with Wrangler. To use a Wasm module in a Worker developed with Wrangler, add the following to your Worker:
```
```suggestion
This is also the basis of Wasm support with Wrangler. To use a Wasm module in a Worker developed with Wrangler, add the following to your Worker:
```

deadlypants1973 reviewed (changes requested):
PCX review

KianNH commented:
I *believe* dynamic imports (i.e `await import`) are now supported in Workers, as teased in https://twitter.com/jasnell/status/1546498745054814209

I don't recall seeing it in changelogs but a few people reported it now works for them. If that's the case, might be worth mentioning?

rozenmd commented on content/workers/wrangler/bundling.md:
```diff
@@ -0,0 +1,48 @@
+# Wrangler's bundling
+
+{{<Aside type="note">}}
+We recommend using Wrangler's inbuilt bundling, but we understand there are cases where you'll need a bit more flexibility. We've built an escape hatch in the form of [Custom Builds](/workers/wrangler/custom-builds/), which let you replace Wrangler's built-in build step with your own.
```
It's a programming thing: https://wiki.c2.com/?EscapeHatch

penalosa commented:
@deadlypants1973 I've addressed your comments, but I've also added an additional section that's probably worth reviewing. In terms of the link to the custom builds page, I think it's worth keeping for right now (I've added a quick link to the custom builds configuration to that page) as we'll hopefully be expanding that page soon.
@rozenmd I've added an extra section mentioning the downsides of `--no-bundle`; no D1 support and that the dashboard might blow up with errors (more diplomatically worded :joy:)—was there anything else you wanted added here?

rozenmd reviewed (approved):
This looks great to me, I'll let @deadlypants1973 have a final review though

deadlypants1973 commented on content/workers/wrangler/custom-builds.md:
```diff
@@ -8,12 +8,12 @@ weight: 4
 {{<Aside type="note">}}
 
-This page is in progress.
+This page is in progress. Refer to the [configuration](/workers/wrangler/configuration/#custom-builds) for a brief overview.
```
```suggestion
This page is in progress. Refer to [Configuration](/workers/wrangler/configuration/#custom-builds) for a brief overview.
```

GregBrimble commented on content/workers/wrangler/bundling.md:
```diff
@@ -0,0 +1,62 @@
+By default, Wrangler bundles your Worker code using [`esbuild`](https://esbuild.github.io/). This means that Wrangler has built-in support for importing modules from [npm](https://www.npmjs.com/) defined in your `package.json`. To review the exact code that Wrangler will upload to Cloudflare, run `wrangler publish --dry-run --outdir dist`, which will show your Worker code after Wrangler's bundlin ...
+
+{{<Aside type="note">}}
+We recommend using Wrangler's inbuilt bundling, but we understand there are cases where you will need more flexibility. We have built an escape hatch in the form of [Custom Builds](/workers/wrangler/custom-builds/), which let you run your own build before Wrangler's built-in one.
```
```suggestion
We recommend using Wrangler's inbuilt bundling, but we understand there are cases where you will need more flexibility. We have built an escape hatch in the form of [Custom Builds](/workers/wrangler/custom-builds/), which lets you run your own build before Wrangler's built-in one.
```

GregBrimble replied:
idk, maybe not.

GregBrimble commented on content/workers/wrangler/bundling.md:
```diff
@@ -0,0 +1,62 @@
+
+## Files which will not be bundled
+
+Bundling your Worker code takes multiple modules and bundles them into one. Sometimes, you might have modules that should not be inlined directly into the bundle. For example, instead of bundling a WASM file into your JavaScript Worker, you would want to upload the WASM file as a separate module that can be imported at runtime. Wrangler supports this for the following file types:
```
```suggestion
Bundling your Worker code takes multiple modules and bundles them into one. Sometimes, you might have modules that should not be inlined directly into the bundle. For example, instead of bundling a Wasm file into your JavaScript Worker, you would want to upload the Wasm file as a separate module that can be imported at runtime. Wrangler supports this for the following file types:
```

deadlypants1973 reviewed (approved):
PCX review
